### PR TITLE
implementing feedback and initial experiments

### PR DESCRIPTION
**Pull request by tomsilver**
@ywcui1990 @subutai @chetan51 please review

### CODE REVIEW

**Comment by chetan51** [comment]
I'm confused about the naming here. So far we're been using the term `external` to refer to inputs on the distal dendrites that are coming from external sources (ie. not lateral connections). Now it seems you're adding a bunch of logic that refers to `external`, where in fact in think you mean `feedback`. Can you clarify a bit on what you mean with the naming?

**Comment by ywcui1990** [comment]
I would suggest to use the term 'apical' or 'feedback' to refer to the feedback inputs.

**Comment by tomsilver** [comment]
Ah, I thought that external just meant "not lateral", so I thought apical
was a type of external. The previous implementation makes more sense now. I
will make the naming change to "apical" to be consistent.

**Comment by chetan51** [comment]
We can go with (`distal` and `apical`) or (`predictive` and `feedback`) connections. I personally prefer the latter, since it's more descriptive, but it's not a strong preference.

**Comment by tomsilver** [comment]
`predictive` is misleading I think because feedback also results in
predictions. i'm going to go with `distal` and `apical`

**Comment by subutai** [comment]
I think apical is more accurate at this point. Some of the learning rules we have/will implement will be tied to the neuroscience of apical dendrites.

**Comment by cogmission** [comment]
distal usually is related to proximal isn't it?

**Comment by cogmission** [comment]
and apical usually is opposed to basal, no?

**Comment by tomsilver** [comment]
changed naming scheme to distal/apical @chetan51 @ywcui1990 @subutai

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 1]
```diff
@@ -147,10 +181,12 @@ def computeFn(self,
     @return (tuple) Contains:
                       `activeCells`               (set),
                       `winnerCells`               (set),
-                      `activeSegments`            (set),
+                      `activeDistalSegments`      (set),
+                      `activeApicalSegments`      (set),
                       `predictiveCells`           (set),
                       'predictedColumns'          (set),
-                      'matchingSegments'          (set),
+                      'matchingDistalSegments'    (set),
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: If we're not going to change all the existing variable names to include `distal` (for example, `prevMatchingSegments`), then we shouldn't change any of them. For now, I would leave the existing variable names as-is (assume that anything that isn't `apical` is `distal`), and just add new variables.

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 2]
```diff
@@ -171,45 +207,86 @@ def computeFn(self,
...
+       connections,
+       apicalConnections)
 
     activeCells.update(_activeCells)
     winnerCells.update(_winnerCells)
 
     if learn:
-      prevCellActivity = prevActiveExternalCells
+      self.learnOnExternalSegments(prevActiveApicalSegments,
+                           apicalLearningSegments,
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: Fix alignmnt

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 3]
```diff
@@ -171,45 +207,86 @@ def computeFn(self,
...
-       connections)
+       connections,
+       apicalConnections)
 
     activeCells.update(_activeCells)
     winnerCells.update(_winnerCells)
 
     if learn:
-      prevCellActivity = prevActiveExternalCells
+      self.learnOnExternalSegments(prevActiveApicalSegments,
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: Change function name from `learnOnExternalSegments` to `learnOnApicalSegments`.

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 4]
```diff
@@ -171,45 +207,86 @@ def computeFn(self,
...
-                           winnerCells,
-                           prevCellActivity,
-                           connections,
-                           predictedInactiveCells,
-                           prevMatchingSegments)
+    (activeDistalSegments,
+    predictiveDistalCells,
+    matchingDistalSegments,
+    matchingDistalCells) = self.computePredictiveCells(
+    activeCells | activeExternalCells, connections)
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: Fix alignment

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 5]
```diff
@@ -171,45 +207,86 @@ def computeFn(self,
...
-    (activeSegments,
-     predictiveCells,
-     matchingSegments,
-     matchingCells) = self.computePredictiveCells(
-       activeCells | activeExternalCells, connections)
+    (activeApicalSegments,
+    predictiveApicalCells,
+    matchingApicalSegments,
+    matchingApicalCells) = self.computeExternalPredictiveCells(
+    activeApicalCells, apicalConnections)
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: Alignment

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 6]
```diff
@@ -171,45 +207,86 @@ def computeFn(self,
...
 
-    (activeSegments,
-     predictiveCells,
-     matchingSegments,
-     matchingCells) = self.computePredictiveCells(
-       activeCells | activeExternalCells, connections)
+    (activeApicalSegments,
+    predictiveApicalCells,
+    matchingApicalSegments,
+    matchingApicalCells) = self.computeExternalPredictiveCells(
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: Update function name

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 7]
```diff
@@ -171,45 +207,86 @@ def computeFn(self,
...
+    activeApicalCells, apicalConnections)
+
+    matchingCells = matchingDistalCells | matchingApicalCells
+
+    # There is probably a way to speed up the below
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: Move below logic to its own function

**Comment by chetan51** [comment]
Done with review.

**Comment by tomsilver** [comment]
thanks @chetan51 , I addressed your comments in the latest commit.

**Comment by chetan51** [comment]
@tomsilver Does this pass the existing tests, including the SensorimotorTemporalMemoryTest?

**Comment by tomsilver** [comment]
@chetan51  the latest commit passes existing tests except for those that were already broken  (TemporalPoolerMonitorMixinTest.testGetConfusionMetrics and OneDUniverseTest.testDecodeSensorValue). I had to make a small temporary change to FastGeneralTemporyMemory to get the tests to pass, which I'll go back and flesh out when GeneralTemporalMemory is finished.

**Pull request merged**